### ISSUE TITLE
Add purescript-leibniz for PureScript

### ISSUE THREAD

**Issue by DonaldKellett** [Codewars/codewars-runner-cli#693]
[purescript-leibniz](https://pursuit.purescript.org/packages/purescript-leibniz/5.0.0) is a PureScript package by **paf31** with a single module named `Data.Leibniz`.  Its `Leibniz` data type is used to express type equality in PureScript.  It would be an interesting addition to the current PureScript environment on Codewars; PureScript does not support GADTs by default but the aforementioned library can [emulate it to some degree](http://code.slipthrough.net/2016/08/10/approximating-gadts-in-purescript/) which I suspect *may* open up Haskell proof Katas (e.g. Odd + Even = Odd, A + B = B + A) for translating into PureScript.

**Comment by kazk** [comment]
It's not in [package sets](https://github.com/purescript/package-sets) so you need to [add](https://psc-package.readthedocs.io/en/stable/maintenance.html#add-a-package-to-the-package-set) it first. This will also ensure that it's compatible.

> PureScript does not support GADTs by default but the aforementioned library can [emulate it to some degree](http://code.slipthrough.net/2016/08/10/approximating-gadts-in-purescript/) which I suspect _may_ open up Haskell proof Katas (e.g. Odd + Even = Odd, A + B = B + A) for translating into PureScript.

You should probably experiment locally if it does what you need before doing that though. See [my comment](https://github.com/Codewars/codewars-runner-cli/issues/692#issuecomment-461278996) on how PureScript works on Codewars.

**Comment by DonaldKellett** [comment]
> It's not in [package sets](https://github.com/purescript/package-sets) so you need to [add](https://psc-package.readthedocs.io/en/stable/maintenance.html#add-a-package-to-the-package-set) it first. This will also ensure that it's compatible.

Thanks for the links - I have currently submitted a pull request to `package-sets` and will notify you once it's approved :smile:

I suppose all future packages not currently in `package-sets` could be added this way?

**Comment by kazk** [comment]
> I suppose all future packages not currently in `package-sets` could be added this way?

Yeah, I don't know if they accept any PureScript packages or not, but a package needs to be in the main `package-sets` to be installed with `psc-package` (we don't want to maintain our own package set). So if a package you want is there, you need to add it by following their instructions. They were very quick when I added `spec-discovery`.

Just to be clear, I won't add every package requested. I'll add a package if it's in `package-sets` that's compatible AND if it's generally useful or has clear use cases.

So let me know after confirming that it does what you need:

> I suspect _may_ open up Haskell proof Katas (e.g. Odd + Even = Odd, A + B = B + A) for translating into PureScript.

**Comment by DonaldKellett** [comment]
So I just read the entry for [Haskell/GADT](https://en.wikibooks.org/wiki/Haskell/GADT) on Wikibooks which stated that while GADTs (in Haskell) can be used as "syntactical sugar" for smart constructors using phantom types and the like (essentially the example given [in this article](
http://code.slipthrough.net/2016/08/10/approximating-gadts-in-purescript/) for PureScript), there are some things that *cannot* be done without GADTs.  For example, the following type-level encoding of the natural numbers (used in various Haskell proof Kata such as A + B = B + A) do not seem to be achievable using (only?) phantom types and smart constructors:

```haskell
{-# LANGUAGE GADTs #-}

-- Code example adapted from the Preloaded section of A + B = B + A and related Kata

data Z
data S n

data Natural n where
  NumZ :: Natural Z
  NumS :: Natural n -> Natural (S n)
```

Indeed, I have experimented with PureScript locally in the past hour or two and managed to get `NumZ` working with the help of `Data.Leibniz` (which I suspect may *not* even be necessary after reading the section on phantom types and smart constructors) but couldn't seem to get `NumS` to work no matter what I tried.  As for other type-level encodings of the natural numbers (which *do* exist in PureScript), none of them seem to be constructed in a manner suitable for proof-related Kata as far as I know, at least not in an easily "katable" way.

> Just to be clear, I won't add every package requested. I'll add a package if it's in `package-sets` that's compatible AND if it's generally useful or has clear use cases.

I guess we should close this Issue then?  At least until I figure out a way to translate the aforementioned Haskell Kata into PureScript (which may or may not even require `Data.Leibniz`) ...
